stdlib/asm/crypto/hashes/sha2_512.masm

```diff
@@ -0,0 +1,88 @@
+use std::mem
+use std::sys
+use std::word
+use std::crypto::hashes::rpo
+
+# ENCODING CONVENTIONS
+# ================================================================================================
+#
+# - Input bytes are stored in memory as packed u32 values in little-endian format
+# - DIGEST_U32[16] = [d_0, ..., d_15] with d_i = u32::from_le_bytes([b_{4i}, ..., b_{4i+3}])
+# - SHA2-512 digest is represented as 16 u32 limbs
+
+# CONSTANTS
+# ================================================================================================
+
+const SHA2_512_HASH_BYTES_EVENT = event("stdlib::hash::sha2::sha512::hash_bytes")
+
+# PROCEDURES
+# ================================================================================================
+
+#! Computes SHA2-512 hash of data stored in memory.
+#!
+#! Input:  [ptr, len_bytes, ...]
+#! Output: [DIGEST_U32[16], ...]
+pub proc hash_bytes
+    exec.hash_bytes_impl
+    # => [COMM, TAG, DIGEST_U32[16], ...]
+
+    exec.sys::log_precompile_request
+    # => [DIGEST_U32[16], ...]
+end
+
+# INTERNAL PROCEDURES
+# ================================================================================================
+
+#! Internal implementation of memory-based SHA2-512 computation via deferred computation.
+#!
+#! Emits an event to populate the advice stack with the digest and store the preimage for deferred
+#! verification. Returns the commitment to the calldata and its tag.
+#!
+#! Input:  [ptr, len_bytes, ...]
+#! Output: [COMM, TAG, DIGEST_U32[16], ...]
+@locals(24)
+pub proc hash_bytes_impl
+    emit.SHA2_512_HASH_BYTES_EVENT
+    # => [ptr, len_bytes, ...]
+
+    # Compute len_u32 = ceil(len_bytes / 4)
+    dup.1 add.3 u32shr.2 swap
+    # => [ptr, len_u32, len_bytes, ...]
+
+    # Load digest from advice stack (16 u32 limbs) and write to digest_ptr = locaddr.0
+    # simultaneously compute COMM_DIGEST = Rpo(DIGEST[0..16])
+    locaddr.0
+    exec.rpo::init_no_padding
+    # => [ZERO, ZERO, ZERO, digest_ptr, ptr, len_u32, len_bytes, ...]
+    repeat.2
+        adv_pipe
+        hperm
+    end
+    exec.rpo::squeeze_digest
+    # => [COMM_DIGEST, digest_ptr, ptr, len_u32, len_bytes, ...]
+
+    movup.4 drop
+    # => [COMM_DIGEST, ptr, len_u32, len_bytes, ...]
+
+    # prepare stack to compute COMM_INPUT = rpo::hash_elements(ptr, len_u32)
+    movup.5 movup.5 exec.rpo::hash_elements
+    # => [COMM_INPUT, COMM_DIGEST, len_bytes, ..]
+
+    # Merge COMM_INPUT and COMM_DIGEST to compute COMM = RPO(COMM_INPUT | COMM_DIGEST)
+    swapw hmerge
```

Comment:
Similar to the above comment, let's use `exec.rpo::merge` instead `hmerge` here.